tgg/relatedness/n_non_ref_pairs.py

```diff
@@ -0,0 +1,273 @@
+        logger.info("Removing AS_lowqual sites...")
+        info_ht = hl.read_table(info_ht_path(*tranche_data, split=True))
+        mt = mt.filter_rows(~info_ht[mt.row_key].AS_lowqual)
+
+    if interval_qc_regions:
+        logger.info("Filtering to interval QC pass regions...")
+        interval_ht = hl.read_table(interval_qc_path(*tranche_data, "autosomes"))
+        mt = mt.filter_rows(hl.is_defined(interval_ht[mt.locus]))
+
+    if adj_only:
+        logger.info("Filtering to adj and calculating allele count...")
+        mt = filter_to_adj(mt)
+
+    logger.info("Changing old callstat annotation...")
+    mt = mt.transmute_rows(original_call_stats=mt.call_stats)
+
+    logger.info("Recalculating call stats post-quick QC...")
+    mt = mt.annotate_rows(hail_call_stats=hl.agg.call_stats(mt.GT, mt.alleles))
+
+    logger.info(
+        "Get AC at alt allele(call_stats includes a count for each allele, including reference) and annotate with hom_var, het, hom_alt..."
+    )
+    mt = mt.annotate_rows(
+        ac=mt.hail_call_stats.AC[1],
+        n_het=hl.agg.count_where(hl.is_defined(mt["GT"]))
+        - hl.sum(mt.hail_call_stats.homozygote_count),
+        n_hom_ref=mt.hail_call_stats.homozygote_count[0],
+        n_hom_alt=mt.hail_call_stats.homozygote_count[1],
+    )
+    mt = mt.annotate_rows(
+        n_non_ref=hl.agg.count_where(hl.is_defined(mt["GT"])) - mt.n_hom_ref
+    )
+
+    logger.info("Filtering sites to where n_non_ref = %i...", non_ref_samples)
+    mt = mt.filter_rows(mt.n_non_ref == non_ref_samples)
+
+    if het_only:
+        mt = mt.filter_rows(mt.n_hom_alt == 0)
+
+    ht = mt.rows().checkpoint(
+        f"{temp_path}/n_non_ref_{non_ref_samples}_sample_high_quality_sites.ht",
+        overwrite=True,
+    )
+    return ht
+
+
+def get_and_count_sample_pairs(
+    mt: hl.MatrixTable, temp_path=TEMP_PATH, non_ref_samples=3,
+) -> hl.Table:
+    """
+    Return the number of shared n non_ref sites per pair.
+
+    :param mt: Matrix Table to compute pairs on.
+    :param temp_path: Path to write pair HT to. Default to TEMP_PATH.
+    :param non_ref_samples: Number of non_ref samples per site. Defaults to 3.
+    :return ht: MatrixTable
+    """
+    logger.info("Collecting samples and counting sample pairs...")
+    mt = mt.annotate_rows(
+        samples=hl.agg.filter(mt.GT.is_non_ref(), hl.agg.collect(mt.s))
+    )
+    mt = mt.annotate_rows(
+        sample_pairs=hl.range(0, mt.samples.length()).flatmap(
+            lambda i: hl.range(i + 1, mt.samples.length()).map(
+                lambda j: hl.tuple([mt.samples[i], mt.samples[j]])
+            )
+        )
+    )
+    ht = mt.select_rows(mt.sample_pairs).rows()
+    ht = ht.explode(ht.sample_pairs)
+
+    logger.info("Aggregating shared sites per pair...")
+    ht = ht.group_by(ht.sample_pairs).aggregate(n_non_ref_sites_shared=hl.agg.count())
+    ht = ht.checkpoint(
+        f"{temp_path}/pairwise_shared_{non_ref_samples}_sites.ht", overwrite=True
+    )
```

Comment:
since this table isn't used downstream, this `checkpoint` could be a `write` (unless you want to print some kind of summary stat beneath this, like "Found %i sample pairs")

Reply:
Good point!